File: leetcode solutions/C++/house robber 2/houseRobber2.cpp
class Solution {
public:
    int rob (vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
            return 0;

        if (n == 1)
            return nums[0];

        return max(robHelper(nums, 0, n-2), robHelper(nums, 1, n-1));
    }

private:
    int robHelper(vector<int>& nums, int start, int end) {
        int pre = 0, cur = 0;

        for (int i = start; i <=end; i++) {
            int value = max(pre + nums[i], cur);
            pre = cur;
            cur = value;
        }
        return cur;
    }
};

File: leetcode solutions/Python/house robber 2/houseRobber2.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0

        n = len(nums)

        if n == 1:
            return nums[0]

        return max(self.robHelper(nums, 0, n - 2), self.robHelper(nums, 1, n -1))


    def robHelper(self, nums, start, end):
        pre = 0
        cur = 0

        for i in xrange(start, end+1):
            value = max(pre + nums[i], cur)
            pre = cur
            cur = value

        return cur